design/design.md

```diff
@@ -0,0 +1,251 @@
+
+## Mathematical rigor and production-readiness
+At this point, it is not our goal to make the implementation 100% correct and
+secure.
+We need to roll out a prototype so that our primary clients (Peras, etc) can
+incorporate it in their prototypes, and also so that the external community can
+start playing with it as well.
+This means that there might be tiny rounding errors in parameter derivation due
+to the use of floats, and there might be tiny errors when using a uniformly
+random hash output to sample some distribution needed for ALBA
+(e.g., Bernoulli, Uniform, Binomial, Poisson).
+We will also probably not rigorously analyze the $H_1$ optimization described in
+section 3.3 of the paper.
+
+## Parameter calculation
+ALBA has two kinds of parameters: user friendly parameters such as
+$\lambda_{sec}$, $\lambda_{rel}$, $n_p$, $n_f$,
+and algorithm specific parameters such as $u$, $d$ and $q$.
+
+Our implementation will have parameter derivation code converting user friendly
+to algorithm specific parameters.
+However, this should be decoupled from the ALBA algorithms:
+through documentation, we will encourage the developer to run parameter
+derivation manually in advance and then hardcode the output algorithm specific
+parameters in their application.
+The ALBA functions such as `prove` and `verify` will thus receive as input
+the algorithm specific parameters.
+
+The rational is that we want to be able to evolve parameter derivation
+(e.g., to optimize parameters for performance, or making the derivation logic
+more rigorous) while maintaining backward compatibility with previous versions
+of the library.
+
+## Deviations from the ALBA paper
+In the future, our implementation will likely deviate from the paper.
+
+One reason is that the definitions given there are quite poor.
+They are, unfortunately, inapplicable for most applications.
+For example, the assumption that there exists a weight oracle, reporting
+the weight of any data element given as input, is often unrealistic.
+In Mithril, for instance, light clients don't know the whole blockchain, but
+would need to verify that signatures in an ALBA certificate come from
+accounts with real stake.
+The ALBA paper simply assumes that the light clients can query the weight oracle
+to get the stake of the signature's creator, but in practice the ALBA
+certificate will need to contain for each signature the correct stake and
+a proof of its correctness (e.g., by supplying a merkle proof).
+Another problem not addressed in the paper is grinding;
+for example, the adversary can move his money to new accounts thus creating
+new valid signatures, which can help him create a rogue ALBA proof.
+These are research problems, and when solved, we will need to enhance / modify
+our implementation as well.
+
+Additionally, we should investigate using the Poisson distribution instead of
+the Binomial as it can potentially simplify calculations and help make our code
+rigorous in the future.
```

Comment:
it might be better to be specific here and explain with a sentence what exactly is not rigorous with our code at the moment?

Reply:
I thought we were not going to do this?